### PR TITLE
refactor!: Updated agent to use require-in-the-middle to register CommonJS instrumentation

### PR DESCRIPTION
**Pull request by bizob2828**
## Description

This PR replaces our instrumentation of Module._load and Module._resolveFilename to get CommonJS instrumentation support.  It is expected to see failing tests for esm versioned tests.  These will no work until #1760 lands. I could skip to get a green CI or consciously ignore, I'd like to get feedback on that part.  This partially supersedes #1759.

There is some nuance to this change.  We have some packages where we either instrumented sub modules or instrumented the root and did a `shim.require` on the sub module.  `require-in-the-middle` supports instrumenting submodules but not both the root and submodules.  In those cases(see amqplib, grpc), you'll see updates to include new module names to directly instrument sub modules.  Personally I think it makes it more clear.  In the case of pino where we instrumented the root then shim.required lib/tools and lib/levels, this no longer works because we get into an infinite loop where lib/levels requires lib/tools.  I was able to work around this by just instrumenting lib/tools directly and getting the level map from the pino instance.  

Most of the work is in shimmer where we store an array of packages to hooks.  Modules are added as we called `shimmer.registerInstrumentation` which occurs within agent and also custom instrumentation.  The reason this is an array and we check if a module exists in array before adding it is because we rely on the reference scoping to add modules after we instantiate the require in the middle class.  This also removes the hook of onResolved because we require-in-the-middle instruments require we can safely instrument all as onRequire, aside from aws-sdk.  You can see the work done https://github.com/newrelic/node-newrelic-aws-sdk/pull/205, which just instruments the smity client send method and only adds middleware once.  

Lastly, a big change was ad-hoc instrumenting of packages.  require-in-the-middle only supports instrumenting modules. Modules live in node_modules unless they are absolute paths. This is a breaking change and I will provide a punch list below that will go into the changelog.  

## Changelog details
Put this into the changelog on release

 * **Breaking change**: Updated agent to use `require-in-the-middle` to register CommonJS instrumentation.
   * Removed `onResolved` hook. If you're using custom instrumentation with an onResolved hook, you must update to use onRequire
   * You can no longer instrument files that are not within a node_module unless you provide an absolute path to the file when registering the instrumentation.
   * You cannot instrument both the base module and a sub module.

### CODE REVIEW

**Review by mrickard — APPROVED** [review]

**Review by jmartin4563 — CHANGES_REQUESTED** [review]
I'm also concerned about how `instrumentLoadedModule` will work with this and the registering of multiple instrumentations. Additionally have we checked that the security agent works with all this stil?

File: test/versioned/express-esm/package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "tests": [
     {
       "engines": {
-        "node": ">=16.12.0 <20"
+        "node": "<=16.12.0"
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply test/versioned/express-esm/package.json:9]: This is done only in this branch to get a green CI. The iitm branch restores this

File: lib/shimmer.js [hunk 2]
```diff
@@ -638,70 +584,6 @@ function _postLoad(agent, nodule, name, resolvedName) {
...
-function _instrumentOnResolved(agent, moduleName, resolvedFilepath) {
-  const instrumentations = shimmer.registeredInstrumentations[moduleName]
-  instrumentations.forEach((instrumentation) => {
-    const shim = shims.createShimFromType(instrumentation.type, agent, moduleName, resolvedFilepath)
-
-    trackInstrumentationUsage(
-      agent,
-      shim,
-      instrumentation.specifier || moduleName,
-      NAMES.FEATURES.INSTRUMENTATION.ON_RESOLVED
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/shimmer.js:676]: Correct I have it on my list to queue up

**Comment by bizob2828** [comment]
> I'm also concerned about how `instrumentLoadedModule` will work with this and the registering of multiple instrumentations. Additionally have we checked that the security agent works with all this still? _ducks to avoid flying objects_

I don't have any of those concerns as those interactions are a layer above. I can double check them though.

File: lib/shimmer.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/shimmer.js:676]: https://source.datanerd.us/agents/angler/pull/476

**Comment by bizob2828** [comment]
Ran security agent on node 18 and 20 both clean runs.  Here's a snippet of output

```sh
===============================================================
PASS

real	4m36.526s
user	11m45.966s
sys	3m13.235s
      276.98 real       705.96 user       193.24 sys
```

**Comment by bizob2828** [comment]
instrumentLoadedModule just works: https://staging.onenr.io/09MR2ozAkjY

```js
const express = require('express')
const newrelic = require('newrelic')
newrelic.instrumentLoadedModule('express', express) 
const app = express()
const { PORT = 3000 } = process.env
const lib = require('@aws-sdk/client-sns')
const sns = new lib.SNSClient()

app.get('/named-route', (req, res) => {
  res.send('hi')
})

app.get('/sns', async (req, res) => {
  const cmd = new lib.ListTopicsCommand({})
  const result = await sns.send(cmd)
  res.send(result)

})

app.listen(PORT, () => console.log(`Example app listening on port ${PORT}!`))

````

I'm going to check ESM app because i have a feeling this will now work too

**Review by jmartin4563 — APPROVED** [review]

**Pull request merged**